src/parser/uniterm.rs

```diff
@@ -141,3 +307,105 @@ impl TryFrom<UniRecord> for Types {
         }
     }
 }
+
+/// Post-process a type at the right hand side of an annotation by replacing each unbound type
+/// variable by a term variable with the same identifier seen as a custom contract
+/// (`AbsType::Var(id)` to `AbsType::Flat(Term::Var(id))`).
+///
+/// Since parsing is done bottom-up, and given the specification of the uniterm syntax for
+/// variables occurring in types, we often can't know right away if a such a variable occurrence
+/// will eventually be a type variable or a term variable seen as a custom contract.
+///
+/// Take for example `a -> b`. At this stage, `a` and `b` could be both variables referring to a
+/// contract (e.g. in `x | a -> b`) or a type variable (e.g. in `x | forall a b. a -> b`),
+/// depending on enclosing `forall`s. To handle both cases, we initially parse all variables inside
+/// types as type variables. When reaching the right-hand side of an annotation, because `forall`s
+/// can only bind locally in a type, we can then decide the actual nature of each occurrence. We
+/// thus recurse into the newly constructed type to change those type variables that are not
+/// actually bound by a `forall` to be term variables. This is the role of `fix_type_vars()`.
+///
+/// Once again because `forall`s only bind variables locally, and don't bind inside contracts, we
+/// don't have to recurse into contracts and this pass will only visit each node of the AST at most
+/// once in total.
+///
+/// There is one subtlety with unirecords, though. A unirecord can still be in interpreted as a
+/// record type later. Take the following example:
+///
+/// ```
+/// let mk_pair : forall a b. a -> b -> {fst: a, snd: b} = <exp>
+/// ```
+///
+/// Since this unirecord will eventually be interpreted as a record type, we can't know yet when
+/// parsing `fst: a` if `a` will be a type variable or a term variable (while, for all other
+/// constructs, an annotation is a boundary that `forall` binders can't cross). In this example,
+/// there is indeed an enclosing forall binding `a`. With unirecords, before fixing type variables,
+/// we have wait until we eventually convert the unirecord to a term (in which case we fix all the
+/// top-level annotations) or a type (in which case we do nothing: the enclosing type will trigger
+/// the fix once it's fully constructed). Fixing a unirecord prior to a conversion to a term is
+/// done by [`fix_fields_types`].
+pub fn fix_type_vars(ty: &mut Types) {
+    fn fix_type_vars_aux(ty: &mut Types, mut bound_vars: Cow<HashSet<Ident>>) {
```

Comment:
Ah, yes, it is annoying to represent an environment in Rust (by which I mean a set or a map, or even a list which is updated locally).

Are there documented best practices? My guess would be that the best way is to implement the local update by a mutation and the inverse mutation on the way out, which I find tricky to get right (and precludes parallelisation, but that's a bit of a minor concern here).

Reply:
I'm guessing, with esprit d'escalier, that one could rely on a purely `Rc`-d data structure (Or `Arc`-d, which makes it a touch tricky to implement such a data structure, if it always has to be implemented twice).